### ISSUE TITLE
Deadlock with running Java application

### ISSUE THREAD

**Issue by jangjae** [s5z/zsim#172]
Hi,
I'm trying to run Java applications on zsim.
I tried some simple Java programs like helloworld, and they work well.
But, with more complex programs like DaCapo benchmark, 
it ends with deadlock with following messages.

[Test system]
Ubuntu 12.04 (3.13 kernel) / Pin-2.14-71313
JDK: 1.7.0_151 / DaCapo ver: 9.12
config file: het.cfg with only beefy cores (using 8 cores)

> ...
> [S 0] Detected possible stall due to fake leaves (1 current)
> [S 0]  [0/26] futex (202) @ 0x7fffe2eb3f1a
> [S 0] Blacklisting from future fake leaves: [0] futex @ 0x7fffe2eb3f1a | arg0 0x7fffdc308f28 arg1 0x80
> [S 0] WARN: [4] ContextChange, reason SIGNAL, inSyscall 0
> [S 0] WARN: [4] ContextChange, reason SIGRETURN, inSyscall 1
> [S 0] Thread 29 starting
> [S 0] [29] Post-patching SYS_sched_getaffinity size 32 cpuset 0x7fff600008c0
> [S 0] [29] Post-patching SYS_sched_getaffinity size 32 cpuset 0x7fff600008c0
> [S 0] Thread 30 starting
> [S 0] [30] Post-patching SYS_sched_getaffinity size 32 cpuset 0x7fff540008c0
> [S 0] [30] Post-patching SYS_sched_getaffinity size 32 cpuset 0x7fff540008c0
> [S 0] Thread 31 starting
> [S 0] [31] Post-patching SYS_sched_getaffinity size 32 cpuset 0x7fff580008c0
> [S 0] [31] Post-patching SYS_sched_getaffinity size 32 cpuset 0x7fff580008c0
> [S 0] Thread 32 starting
> [S 0] [32] Post-patching SYS_sched_getaffinity size 32 cpuset 0x7fff4c0008c0
> [S 0] [32] Post-patching SYS_sched_getaffinity size 32 cpuset 0x7fff4c0008c0
> [S 0] Thread 33 starting
> [S 0] [33] Post-patching SYS_sched_getaffinity size 32 cpuset 0x7fff500008c0
> [S 0] [33] Post-patching SYS_sched_getaffinity size 32 cpuset 0x7fff500008c0
> [S 0] Thread 34 starting
> [S 0] [34] Post-patching SYS_sched_getaffinity size 32 cpuset 0x7fff440008c0
> [S 0] [34] Post-patching SYS_sched_getaffinity size 32 cpuset 0x7fff440008c0
> [S 0] Thread 35 starting
> [S 0] [35] Post-patching SYS_sched_getaffinity size 32 cpuset 0x7fff3c0008c0
> [S 0] [35] Post-patching SYS_sched_getaffinity size 32 cpuset 0x7fff3c0008c0
> [S 0] Thread 36 starting
> [S 0] [36] Post-patching SYS_sched_getaffinity size 32 cpuset 0x7fff400008c0
> [S 0] [36] Post-patching SYS_sched_getaffinity size 32 cpuset 0x7fff400008c0
> [S 0] WARN: [21] Executing unvirtualized potentially timing-sensitive syscall: times (100)
> [GC[S 0] Time slice ended, context-switched 1 threads, runQueue size 0, available 7
> [S 0]  State:  10r  5r  9r 19r 16r 12r 20r  6r
> [S 0] WARN: Watchdog sleeping for a long time due to long sleep, 59647 secs
> [H] WARN: Stalled for 20 secs so far
> [H] WARN: Stalled for 30 secs so far
> [H] WARN: Stalled for 40 secs so far
> [H] WARN: Stalled for 50 secs so far
> [H] WARN: Stalled for 60 secs so far
> [H] WARN: Stalled for 70 secs so far
> [H] WARN: Stalled for 80 secs so far
> [H] WARN: Stalled for 90 secs so far
> [H] WARN: Stalled for 100 secs so far
> [H] WARN: Stalled for 110 secs so far
> [H] WARN: Stalled for 120 secs so far
> [H] WARN: Stalled for 130 secs so far
> [H] WARN: Deadlock detected, killing children
> [H] Received interrupt
> [H] Attempting graceful termination
> [H] Killing process 4769
> [H] Done sending kill signals
> [H] WARN: Hard death at exit (1 children running), killing the whole process

When I lower schedQuantum (10000 -> 1000), 
but it gets stuck with the same warning and terminated as following,
even successful benchmark pass.

> [S 0] Time slice ended, context-switched 1 threads, runQueue size 0, available 7
> [S 0]  State:  26r 24r 27r 22r 25r 18r 17r 21r
> [S 0] Thread 24 finished
> [S 0] Time slice ended, context-switched 1 threads, runQueue size 0, available 7
> [S 0]  State:  23r 28r 26r 22r 25r 18r 17r 21r
> [S 0] Thread 22 finished
> [S 0] Thread 21 finished
> [S 0] Thread 23 finished
> [S 0] Thread 28 finished
> [S 0] Thread 27 finished
> [S 0] Thread 26 finished
> [S 0] Thread 25 finished
> Normal completion.
> ===== DaCapo 9.12 xalan PASSED in 3190 msec =====
> [S 0] Thread 20 finished
> [S 0] WARN: Watchdog sleeping for a long time due to long sleep, 61723 secs
> [H] WARN: Stalled for 20 secs so far
> [H] WARN: Stalled for 30 secs so far
> [H] WARN: Stalled for 40 secs so far
> [H] WARN: Stalled for 50 secs so far
> [H] WARN: Stalled for 60 secs so far
> [H] WARN: Stalled for 70 secs so far
> [H] WARN: Stalled for 80 secs so far
> [H] WARN: Stalled for 90 secs so far
> [H] WARN: Stalled for 100 secs so far
> [H] WARN: Stalled for 110 secs so far
> [H] WARN: Stalled for 120 secs so far
> [H] WARN: Stalled for 130 secs so far
> [H] WARN: Deadlock detected, killing children
> [H] Received interrupt
> [H] Attempting graceful termination
> [H] Killing process 5446
> [H] Done sending kill signals
> [H] WARN: Hard death at exit (1 children running), killing the whole process tree

I saw the similar problem on #44 for overcommit problem,
but Java is inherently overcommitted due to additional threads except the mutator (GC, JIT).

Does anyone saw this kind of problem?
Any comments or suggestion will be helpful.

Thanks,
Jaey

**Comment by gaomy3832** [comment]
You can try to increase the number of cores in the config file (say, to 64, until there is no context switch log messages) to see if it is an overcommit problem similar to #44. If the problem still happens, then it is not related.

The warnings
```[S 0] WARN: [21] Executing unvirtualized potentially timing-sensitive syscall: times (100)```
and
```[S 0] WARN: Watchdog sleeping for a long time due to long sleep, 59647 secs```
look suspicious. Probably the timing is messed up due to the mishandle of the syscall and results in no forward progress.

Also, you may want to try disable the watchdog (set `sim.deadlockDetection` to false in config file) to see if this is a false alarm.

**Comment by jangjae** [comment]
I tried 1) increase # of cores (over-provisioning cores), 
and 2) turn off deadlock detection as your comments.

After applying for both approaches
the same warning on watchdog long sleep still occurs
(times syscall error is disappeared after turning off print GC log option in JVM) 
and stuck at the end of the program without termination.

As you mentioned,
this problem seems not to be overcommit problem.
According to  zsim log, some threads not finished successfully including main thread 0 and stuck at the end.
(total 28 threads are invoked, but only 9 threads are finished)

If this is timing problem, what will be the steps to fix this problem?
FYI, I attached full log files.

[zsim-xalan-64cores.txt](https://github.com/s5z/zsim/files/1316774/zsim-xalan-over.txt)
[zsim-xalan-8cores.txt](https://github.com/s5z/zsim/files/1316798/zsim-xalan-8core.txt)

Jaey

**Comment by jangjae** [comment]
FYI, I used stanford-mast/zsim branch to avoid waitUntilQueued issue.

**Comment by gaomy3832** [comment]
This actually reminds me something, that once someone was trying to run Python on zsim, and he got the same issue of deadlock at the end of the simulation. I would guess that the Java/Python runtime has some internal long-run threads created and never terminated (maybe detached?), so zsim can never wait until all created threads finish. See https://github.com/s5z/zsim/blob/fb4d6e0475a25cffd23f0687ede2d43d96b4a99f/src/zsim.cpp#L1109 So a patch is needed to fix this issue by, for example, identifying these "detached" threads and removing them out of zsim's tracking. Actually, as long as the simulation finishes for the code region of interest, the stats results should be good. So feel free to use some hack for now.

As you said the timing problem is only related to the GC log, so it should be not critical. To fix it, you will have to add new virtualization similar to the other timing syscalls. See src/virt/time.cpp. Currently the warning is from https://github.com/s5z/zsim/blob/fb4d6e0475a25cffd23f0687ede2d43d96b4a99f/src/virt/virt.cpp#L64

**Comment by jangjae** [comment]
This is resolved with 1) changing jdk ver. (works well with jdk7u25 below) and 2) sufficient warm-up runs with fast-forwarding (maybe JIT and code cache warmup problems).
Detailed symptom I checked is that when futex_wait syscall is called, 
its timeout argument acquisition from "pthread_cond_timedwait" argument is set with extremely large value (>10000 secs), which incurs deadlock.
When I compared jdks source codes, problematic parts in jdk should be "park()" codes
and only difference between jdks (7u25 and upper jdk ver.) is the way of timeout value calculation.
Just using under jdk7u25 version can be quick solution for this,
but I think special treat for this in zsim will be needed in the future.
